plugins/bigquery/dbt/adapters/bigquery/impl.py

```diff
@@ -430,17 +431,73 @@ def is_replaceable(self, relation, conf_partition, conf_cluster):
         except google.cloud.exceptions.NotFound:
             return True
 
-        table_partition = table.time_partitioning
-        if table_partition is not None:
-            table_partition = table_partition.field
-
         table_cluster = table.clustering_fields
 
         if isinstance(conf_cluster, str):
             conf_cluster = [conf_cluster]
 
-        return table_partition == conf_partition \
-            and table_cluster == conf_cluster
+        time_partition = table.time_partitioning
+        range_partition = table.range_partitioning
+
+        if time_partition is not None:
+            return time_partition.field == conf_partition.get('field', None) \
+                and table_cluster == conf_cluster
+
+        if range_partition is not None:
+            table_range = range_partition.range_
+            conf_range = conf_partition.get('range', None)
+            return range_partition.field == conf_partition.get('field', None) \
+                and table_range.start == conf_range.get('start', None) \
+                and table_range.end == conf_range.get('end', None) \
+                and table_range.interval == conf_range.get('interval', None) \
+                and table_cluster == conf_cluster
+
+    @available
+    def parse_partition_by(self, raw_partition_by):
+        """
+        dbt v0.16.0 expects `partition_by` to be a dictionary where previously
+        it was a string. Check the type of `partition_by`, raise error
+        or warning if string, and attempt to convert to dict.
+        """
+        if isinstance(raw_partition_by, dict):
+            if raw_partition_by.get('field', None) is not None:
+                if raw_partition_by.get('data_type', None) is not None:
+                    return raw_partition_by
+                else:  # assume date type as default
+                    return dict(raw_partition_by, data_type='date')
+            else:
+                dbt.exceptions.CompilerException(
+                    'Config `partition_by` is missing required item `field`'
+                )
+
+        elif isinstance(raw_partition_by, str):
+            if 'range_bucket' in raw_partition_by.lower():
+                dbt.exceptions.CompilerException('''
+                    BigQuery integer range partitioning (beta) is supported \
+                    by the new `partition_by` config, which accepts a \
```

Comment:
BigQuery integer range partitioning (beta) is only supported
with the dictionary format of the `partition_by` config.

More information: [Link from above]


^ I like using modifiers like `only`, as they help clarify the nature of these validation errors. These error messages should:
- Explain what went wrong
- Explain how to fix it

Also, this error message will invariably stick around for longer than we intend, so I try to avoid temporal words like "new", as things like that tend to cause chuckles a couple of years down the line :p